### PR TITLE
Replace `x != undefined` with `x != null`

### PR DESCRIPTION
**Pull request by j-f1**
~Bonus: replace `undefined == null` and `null != null` with `!0` and `!1`, respectively.~

Tests included.

### CODE REVIEW

File: packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+jest.autoMockOff();
+
+const babel = require("babel-core");
+const plugin = require("../src/index");
+const unpad = require("../../../utils/unpad");
```
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: unused.

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js [hunk 2]
```diff
@@ -9,6 +14,17 @@ module.exports = function() {
       BinaryExpression(path) {
         const { node } = path;
         const op = node.operator;
+        const negated = node.operator.startsWith("!");
+ 
+        if (["!=", "=="].includes(node.operator)) {
```
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: use indexOf, we still support node 4

File: packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js (hunk 1 above)
Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: Done.

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js (hunk 2 above)
Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: Fixed.

File: packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    const expected = "x == null;";
+    expect(transform(source)).toBe(expected);
+  });
+  it("should not shorten `x === undefined`", () => {
+    const source = "x === undefined;";
+    expect(transform(source)).toBe(source);
+  });
+
+  it("should shorten `undefined == null`", () => {
+    const source = "undefined == null";
```
Comment (boopathi) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: this should be already taken care of by constant-folding plugin. This plugin should not worry about these folding constants.

Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: It [doesn’t appear to handle this](http://babeljs.io/repl/#?babili=true&evaluate=false&lineWrap=false&presets=react%2Cstage-0&targets=&browsers=&builtIns=false&debug=false&experimental=false&loose=false&spec=false&code=f(undefined%20%3D%3D%20null)&playground=true). Are you saying that I should move it there?

Comment (boopathi) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: That can be fixed in `path.evaluate` of babel if that doesn't handle it. Or it's an ordering issue of who runs first - #61 #422.

Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: I could add `void 0` to `isNull`.

Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: How does that sound @boopathi?

Comment (boopathi) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: I meant this entire replacement shouldn't be a part of this plugin. It should be handled in `path.evaluate` of babel.

Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/__tests__/eqnull-test.js]: @boopathi Done. Do you know where the source for `path.evaluate` is so I can open a PR there?

**Comment by xtuc** [comment]
I'm wondering if we can do this, isn't it unsafe? `null` and `undefined` are not equal.

**Comment by j-f1** [comment]
@xtuc `null` and `undefined ` are double-equals equal, not triple-equals equal.

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js [hunk 4]
```diff
@@ -9,6 +9,13 @@ module.exports = function() {
       BinaryExpression(path) {
         const { node } = path;
         const op = node.operator;
+
+        if (["!=", "=="].indexOf(node.operator) !== -1) {
+          if (t.isIdentifier(node.right, { name: "undefined" })) {
```
Comment (boopathi) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: `void 0` will not be transformed.

**Comment by boopathi** [comment]
Can you add some tests for this ?

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js (hunk 4 above)
Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: Fixed.

**Comment by j-f1** [comment]
@boopathi Are you assuming in babili that people won’t do `void someFunctionWithSideEffects()`?

**Comment by j-f1** [comment]
@boopathi Tests restored.

**Comment by boopathi** [comment]
> @boopathi Are you assuming in babili that people won’t do `void someFunctionWithSideEffects()`?

Constant-folding plugin should take care of folding constants. `void foo()` should be a deopt anyway here. It should check only for `void x` where `pathOfX.isPure()` is `true`.

**Comment by j-f1** [comment]
@boopathi This is ready for review.

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js [hunk 5]
```diff
@@ -9,6 +15,16 @@ module.exports = function() {
       BinaryExpression(path) {
         const { node } = path;
         const op = node.operator;
+
+        if (["!=", "=="].indexOf(node.operator) !== -1) {
+          if (
+            t.isIdentifier(node.right, { name: "undefined" }) ||
+            isPureVoid(path.get("right"))
```
Comment (boopathi) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: This will work only for `void right`. What about `undefined == foo` ?

Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: @boopathi Fixed.

**Comment by boopathi** [comment]
Just a few thoughts on this -

1. Should it go into comparison-operators?
1. Since this is an unsafe transformation, it's better to put this entire transformation behind an option - `undefinedToNull` or `voidToNull` or anything else relevant. 
1. Update tests to use `thePlugin`
1. `t.isIdentifier("undefined")` check should also check for bindings in the scope `path.scope.getBinding("undefined")` there is a defined `undefined`.

**Comment by j-f1** [comment]
@boopathi 

1. 👍
2. I don’t think it’s unsafe. Why do you think it’s unsafe?
3. 👍
4. 👍

**Comment by xtuc** [comment]
2) it's safe since https://github.com/babel/minify/pull/572#issuecomment-307738921 (see [Equality_comparisons_and_sameness](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness))

**Comment by j-f1** [comment]
@boopathi 

1. It’s already in comparison-operators.
2. ❓
3. ✅
4. ✅

**Comment by boopathi** [comment]
My question is whether it should go into comparison operators, or should it go into simplify. IMO, Comparison operators simplifies the operator used without changing the operands.

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js [hunk 6]
```diff
@@ -1,6 +1,25 @@
...
+    return (
+      path.isIdentifier({ name: "undefined" }) &&
+      !path.scope.getBinding("undefined")
+    );
+  }
+
+  function undefinedToNull(path) {
+    if (isRealUndefined(path) || isPureVoid(path)) {
+      path.replaceWith(t.nullLiteral());
+      return true;
```
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: is this return true / false needed? I don't see it used anywhere for conditional check.

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js [hunk 7]
```diff
@@ -1,6 +1,25 @@
...
-module.exports = function() {
+module.exports = function({ types: t }) {
+  function isPureVoid(path) {
+    return path.isUnaryExpression({ operator: "void" }) && path.isPure();
+  }
+
+  function isRealUndefined(path) {
+    return (
+      path.isIdentifier({ name: "undefined" }) &&
+      !path.scope.getBinding("undefined")
```
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: why is this check required?. Even if someone does 
```js
var undefined = 'blah';
```

Its a read only property and does not impact our null conversion.

Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: @boopathi said:

> `t.isIdentifier("undefined")` check should also check for bindings in the scope `path.scope.getBinding("undefined")` there is a defined `undefined`.

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js (hunk 6 above)
Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: Removed 🔥

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js (hunk 7 above)
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: /cc @boopathi Any reason why the check is needed? Am i missing some edge case?

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js (hunk 6 above)
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: 👍

File: packages/babel-plugin-transform-simplify-comparison-operators/src/index.js (hunk 7 above)
Comment (boopathi) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]:
```js
function foo(undefined) { return undefined == void 0 } 
foo("bar"); // true or false ?
```

Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: `true` even if undefined or both are converted to null

Comment (j-f1) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]:
```js
> function foo(undefined) { return undefined == void 0 } 
> foo("bar"); // true or false ?
// false
```

Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-simplify-comparison-operators/src/index.js]: okay got it 👍

**Comment by boopathi** [comment]
It's not immediately obvious that this goes into `comparison-operators` plugin. This question still applies - 

> My question is whether it should go into comparison operators, or should it go into simplify. IMO, Comparison operators simplifies the operator used without changing the operands

**Comment by j-f1** [comment]
@boopathi If you’d like me to move it to `simplify`, I’ll gladly do so. Should I?

**Comment by j-f1** [comment]
Sorry for the delay on this @boopathi 😢

I’ve moved the changes to `babel-plugin-minify-simplify`.

**Comment by j-f1** [comment]
Looks like the Travis failure is unrelated. Closing and reopening to restart the build.

**Comment by j-f1** [comment]
Ping @boopathi and @vigneshshanmugam for 👀

**Comment by boopathi** [comment]
@j-f1 Hi, Thanks for the PR. Right now let's concentrate on fixing the existing bugs, babel-7, improving the speed of transformations and fixing the memory issue to get it to 1.0. New set of features may delay that. I'll leave it open and postpone it to sometime later.

**Pull request merged**